File: skills/agents-orchestrator/tests/fixtures/fake_acp_agent.py
#!/usr/bin/env python3
"""Deterministic test Agent implemented on the official ACP Python SDK."""

import argparse
import asyncio
import json
import os
import pathlib
import subprocess
import sys

SCRIPTS_DIR = pathlib.Path(__file__).resolve().parents[2] / "scripts"
if str(SCRIPTS_DIR) not in sys.path:
    sys.path.insert(0, str(SCRIPTS_DIR))

from backends.acp.dependencies import activate as activate_managed_sdk

activate_managed_sdk()

from acp import PROTOCOL_VERSION, RequestError, run_agent
from acp.schema import (
    AgentCapabilities,
    AgentAuthCapabilities,
    AllowedOutcome,
    CloseSessionResponse,
    AgentMessageChunk,
    Implementation,
    InitializeResponse,
    LoadSessionResponse,
    NewSessionResponse,
    PermissionOption,
    PromptResponse,
    SessionCapabilities,
    SessionCloseCapabilities,
    TextContentBlock,
    ToolCallLocation,
    ToolCallUpdate,
    UserMessageChunk,
)


def runtime_action(action_type, payload):
    skill_dir = os.environ["AGENT_SWARM_SKILL_DIR"]
    command = [
        sys.executable,
        os.path.join(skill_dir, "scripts", "agent_orchestrator.py"),
        "action",
        "--type",
        action_type,
        "--stdin",
    ]
    completed = subprocess.run(
        command,
        input=json.dumps(payload),
        text=True,
        capture_output=True,
        env=os.environ.copy(),
        check=False,
        timeout=10,
    )
    if completed.returncode != 0:
        raise RuntimeError(
            completed.stderr or completed.stdout or "Runtime action failed"
        )
    return json.loads(completed.stdout)


class FakeAgent:
    def __init__(self, args):
        self.args = args
        self.client = None
        self.cancelled = asyncio.Event()

    def on_connect(self, connection):
        self.client = connection

    async def initialize(
        self,
        protocol_version,
        client_capabilities=None,
        client_info=None,
        **kwargs,
    ):
        await asyncio.sleep(self.args.initialize_delay)
        return InitializeResponse(
            protocol_version=PROTOCOL_VERSION,
            agent_capabilities=AgentCapabilities(
                load_session=self.args.scenario == "history",
                auth=AgentAuthCapabilities(),
                session_capabilities=SessionCapabilities(
                    close=SessionCloseCapabilities()
                ),
            ),
            agent_info=Implementation(
                name="fake-acp", title="Fake ACP", version="1.0.0"
            ),
        )

    async def new_session(
        self, cwd, additional_directories=None, mcp_servers=None, **kwargs
    ):
        await asyncio.sleep(self.args.session_delay)
        return NewSessionResponse(
            session_id="fake-session-%s-%s"
            % (os.environ.get("AGENT_SWARM_ATTEMPT_ID", "standalone"), os.getpid())
        )

    async def prompt(self, session_id, prompt, **kwargs):
        prompt_text = "\n".join(
            block.text for block in prompt if isinstance(block, TextContentBlock)
        )
        if self.args.prompt_file:
            pathlib.Path(self.args.prompt_file).write_text(
                prompt_text, encoding="utf-8"
            )
        if self.args.scenario == "hold":
            await self.cancelled.wait()
            return PromptResponse(stop_reason="cancelled")
        if self.args.scenario == "crash":
            os._exit(23)
        if self.args.scenario == "split":
            await asyncio.to_thread(self._finish_split, prompt_text)
            return PromptResponse(stop_reason="end_turn")
        if self.args.scenario == "finish":
            await asyncio.sleep(0.25)
            await asyncio.to_thread(self._finish_direct)
            return PromptResponse(stop_reason="end_turn")
        if self.args.scenario == "permission":
            if self.client is None:
                raise RuntimeError("fake Agent has no connected Client")
            response = await self.client.request_permission(
                session_id=session_id,
                tool_call=ToolCallUpdate(
                    tool_call_id="fake-call",
                    title="write",
                    kind="edit",
                    locations=[
                        ToolCallLocation(path=os.path.join(os.getcwd(), "inside.txt"))
                    ],
                ),
                options=[
                    PermissionOption(
                        option_id="allow-once",
                        name="Allow once",
                        kind="allow_once",
                    ),
                    PermissionOption(
                        option_id="deny-once",
                        name="Deny once",
                        kind="reject_once",
                    ),
                ],
            )
            if not isinstance(response.outcome, AllowedOutcome):
                raise RuntimeError("permission was denied")
            return PromptResponse(stop_reason="end_turn")
        return PromptResponse(stop_reason="end_turn")

    async def load_session(
        self, cwd, session_id, mcp_servers=None, additional_directories=None, **kwargs
    ):
        if self.args.scenario != "history":
            raise RequestError.method_not_found("session/load")
        if session_id == "missing-session":
            raise RequestError.resource_not_found(session_id)
        if self.client is None:
            raise RuntimeError("fake Agent has no connected Client")
        await self.client.session_update(
            session_id=session_id,
            update=UserMessageChunk.model_validate(
                {
                    "sessionUpdate": "user_message_chunk",
                    "content": {"type": "text", "text": "remembered user message"},
                    "messageId": "history-user",
                }
            ),
        )
        await self.client.session_update(
            session_id=session_id,
            update=AgentMessageChunk.model_validate(
                {
                    "sessionUpdate": "agent_message_chunk",
                    "content": {"type": "text", "text": "remembered agent response"},
                    "messageId": "history-agent",
                }
            ),
        )
        return LoadSessionResponse()

    async def cancel(self, session_id, **kwargs):
        self.cancelled.set()

    async def close_session(self, session_id, **kwargs):
        return CloseSessionResponse()

    async def ext_method(self, method, params):
        raise RequestError.method_not_found("_" + method)

    async def ext_notification(self, method, params):
        raise RequestError.method_not_found("_" + method)

    def _submit_estimate(self, strategy, complexity, reason):
        return runtime_action(
            "submit_estimate",
            {
                "revision": False,
                "strategy": strategy,
                "resolved_intent": "implement",
                "complexity": complexity,
                "concerns": [],
                "unknowns": [],
                "estimated_files": [],
                "reason": reason,
            },
        )

    def _finish(self, summary, integration=None):
        return runtime_action(
            "finish",
            {
                "status": "done",
                "retryable": False,
                "summary": summary,
                "changed_files": [],
                "artifacts": [],
                "validation": None,
                "review": None,
                "integration_check": integration,
                "caveats": [],
            },
        )

    def _finish_direct(self):
        self._submit_estimate("direct", "low", "deterministic fake task")
        self._finish("fake ACP child finished")

    def _finish_split(self, prompt_text):
        is_parent = "\nchild goal\n" in prompt_text
        self._submit_estimate(
            "split" if is_parent else "direct",
            "medium" if is_parent else "low",
            "deterministic fake split",
        )
        integration = None
        if is_parent:
            created = runtime_action(
                "create_tasks",
                {
                    "tasks": [
                        {
                            "key": "leaf-a",
                            "goal": "leaf-a",
                            "intent_hint": "implement",
                            "output_contract": "finish leaf a",
                        },
                        {
                            "key": "leaf-b",
                            "goal": "leaf-b",
                            "intent_hint": "implement",
                            "output_contract": "finish leaf b",
                        },
                    ]
                },
            )
            task_ids = [item["task_id"] for item in created["tasks"]]
            waited = runtime_action(
                "wait",
                {
                    "task_ids": task_ids,
                    "condition": "all_done",
                    "listen_seconds": 15,
                },
            )
            if not waited.get("complete"):
                raise RuntimeError("fake split children did not finish")
            integration = {
                "status": "passed",
                "summary": "fake leaves integrated",
            }
        self._finish("fake split task finished", integration)


async def async_main(args):
    if args.counter_file:
        path = pathlib.Path(args.counter_file)
        count = int(path.read_text()) if path.exists() else 0
        path.write_text(str(count + 1))
    await run_agent(FakeAgent(args), use_unstable_protocol=True)
    if args.scenario == "hold":
        while True:
            await asyncio.sleep(1)


def main(argv=None):
    parser = argparse.ArgumentParser()
    parser.add_argument("--scenario", default="basic")
    parser.add_argument("--counter-file")
    parser.add_argument("--prompt-file")
    parser.add_argument("--initialize-delay", type=float, default=0)
    parser.add_argument("--session-delay", type=float, default=0)
    args = parser.parse_args(argv)
    asyncio.run(async_main(args))


if __name__ == "__main__":
    main()

File: skills/agents-orchestrator/tests/fixtures/fake_acp_agent.ts
#!/usr/bin/env bun
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { Readable, Writable } from "node:stream";
import * as acp from "@agentclientprotocol/sdk";

interface Options {
  scenario: string;
  counterFile?: string;
  promptFile?: string;
  initializeDelay: number;
  sessionDelay: number;
}

function parse(): Options {
  const value = (name: string): string | undefined => {
    const index = process.argv.indexOf(name);
    return index >= 0 ? process.argv[index + 1] : undefined;
  };
  return {
    scenario: value("--scenario") ?? "basic",
    counterFile: value("--counter-file"),
    promptFile: value("--prompt-file"),
    initializeDelay: Number(value("--initialize-delay") ?? 0),
    sessionDelay: Number(value("--session-delay") ?? 0),
  };
}

function runtimeAction(type: string, payload: unknown): Record<string, any> {
  const skill = process.env.AGENTS_ORCHESTRATOR_SKILL_DIR ?? process.env.AGENT_SWARM_SKILL_DIR;
  if (!skill) throw new Error("missing Runtime skill directory");
  const result = Bun.spawnSync({
    cmd: [process.execPath, join(skill, "scripts", "agent_orchestrator.ts"), "action", "--type", type, "--stdin"],
    stdin: Buffer.from(JSON.stringify(payload)), stdout: "pipe", stderr: "pipe", timeout: 15_000, env: process.env,
  });
  if (result.exitCode !== 0) throw new Error("Runtime action failed");
  return JSON.parse(result.stdout.toString());
}

class FakeAgent {
  private cancelled = new Map<string, () => void>();
  constructor(private readonly options: Options) {}

  async initialize(): Promise<acp.InitializeResponse> {
    await Bun.sleep(this.options.initializeDelay * 1000);
    return {
      protocolVersion: acp.PROTOCOL_VERSION,
      agentCapabilities: {
        loadSession: this.options.scenario === "history",
        auth: {},
        sessionCapabilities: { close: {} },
      },
      agentInfo: { name: "fake-acp", title: "Fake ACP", version: "1.0.0" },
    };
  }

  async newSession(): Promise<acp.NewSessionResponse> {
    await Bun.sleep(this.options.sessionDelay * 1000);
    return { sessionId: `fake-session-${process.env.AGENT_SWARM_ATTEMPT_ID ?? "standalone"}-${process.pid}` };
  }

  private promptText(params: acp.PromptRequest): string {
    return params.prompt.map((item) => item.type === "text" ? item.text : "").filter(Boolean).join("\n");
  }

  async prompt(params: acp.PromptRequest, client: acp.AgentContext): Promise<acp.PromptResponse> {
    const text = this.promptText(params);
    if (this.options.promptFile) writeFileSync(this.options.promptFile, text, "utf8");
    if (this.options.scenario === "hold") {
      await new Promise<void>((resolveCancelled) => this.cancelled.set(params.sessionId, resolveCancelled));
      return { stopReason: "cancelled" };
    }
    if (this.options.scenario === "crash") process.exit(23);
    if (this.options.scenario === "raw-error") throw new Error("AGENT_RAW_SECRET_SENTINEL");
    if (this.options.scenario === "split") { this.finishSplit(text); return { stopReason: "end_turn" }; }
    if (this.options.scenario === "finish") { await Bun.sleep(250); this.finishDirect(); return { stopReason: "end_turn" }; }
    if (this.options.scenario === "permission") {
      const response = await client.request(acp.methods.client.session.requestPermission, {
        sessionId: params.sessionId,
        toolCall: {
          toolCallId: "fake-call", title: "write", kind: "edit",
          locations: [{ path: join(process.cwd(), "inside.txt") }],
        },
        options: [
          { optionId: "allow-once", name: "Allow once", kind: "allow_once" },
          { optionId: "deny-once", name: "Deny once", kind: "reject_once" },
        ],
      });
      if (response.outcome.outcome !== "selected" || response.outcome.optionId !== "allow-once") throw new Error("permission was denied");
    }
    return { stopReason: "end_turn" };
  }

  async loadSession(params: acp.LoadSessionRequest, client: acp.AgentContext): Promise<acp.LoadSessionResponse> {
    if (this.options.scenario !== "history") throw acp.RequestError.methodNotFound("session/load");
    if (params.sessionId === "missing-session") throw acp.RequestError.resourceNotFound(params.sessionId);
    await client.notify(acp.methods.client.session.update, {
      sessionId: params.sessionId,
      update: { sessionUpdate: "user_message_chunk", content: { type: "text", text: "remembered user message" } },
    });
    await client.notify(acp.methods.client.session.update, {
      sessionId: params.sessionId,
      update: { sessionUpdate: "agent_message_chunk", content: { type: "text", text: "remembered agent response" } },
    });
    return {};
  }

  cancel(sessionId: string): void { this.cancelled.get(sessionId)?.(); this.cancelled.delete(sessionId); }

  private estimate(strategy: string, complexity: string, reason: string): void {
    runtimeAction("submit_estimate", {
      revision: false, strategy, resolved_intent: "implement", complexity,
      concerns: [], unknowns: [], estimated_files: [], reason,
    });
  }

  private finish(summary: string, integration: unknown = null): void {
    runtimeAction("finish", {
      status: "done", retryable: false, summary, changed_files: [], artifacts: [],
      validation: null, review: null, integration_check: integration, caveats: [],
    });
  }

  private finishDirect(): void { this.estimate("direct", "low", "deterministic fake task"); this.finish("fake ACP child finished"); }

  private finishSplit(prompt: string): void {
    const parent = prompt.includes("\nchild goal\n");
    this.estimate(parent ? "split" : "direct", parent ? "medium" : "low", "deterministic fake split");
    let integration: unknown = null;
    if (parent) {
      const created = runtimeAction("create_tasks", { tasks: [
        { key: "leaf-a", goal: "leaf-a", intent_hint: "implement", output_contract: "finish leaf a" },
        { key: "leaf-b", goal: "leaf-b", intent_hint: "implement", output_contract: "finish leaf b" },
      ] });
      const waited = runtimeAction("wait", {
        task_ids: created.tasks.map((item: Record<string, any>) => item.task_id), condition: "all_done", listen_seconds: 15,
      });
      if (!waited.complete) throw new Error("fake split children did not finish");
      integration = { status: "passed", summary: "fake leaves integrated" };
    }
    this.finish("fake split task finished", integration);
  }
}

const options = parse();
if (options.counterFile) {
  const count = existsSync(options.counterFile) ? Number(readFileSync(options.counterFile, "utf8")) : 0;
  writeFileSync(options.counterFile, String(count + 1));
}
const fake = new FakeAgent(options);
const stream = acp.ndJsonStream(
  Writable.toWeb(process.stdout) as WritableStream<Uint8Array>,
  Readable.toWeb(process.stdin) as ReadableStream<Uint8Array>,
);
acp.agent({ name: "fake-acp" })
  .onRequest(acp.methods.agent.initialize, () => fake.initialize())
  .onRequest(acp.methods.agent.session.new, () => fake.newSession())
  .onRequest(acp.methods.agent.session.prompt, ({ params, client }) => fake.prompt(params, client))
  .onRequest(acp.methods.agent.session.load, ({ params, client }) => fake.loadSession(params, client))
  .onRequest(acp.methods.agent.session.close, () => ({}))
  .onNotification(acp.methods.agent.session.cancel, ({ params }) => { fake.cancel(params.sessionId); })
  .connect(stream);
